_posts/2019-04-23-autoregressive-flows/experimenting-with-autoregressive-flows.Rmd

```diff
@@ -0,0 +1,645 @@
+df <- read_table2("ethylene_CO.txt",
+                  skip = 1,
+                  col_names = FALSE)
+
+# we don't know if we'll end up with the same columns as the authors did,
+# but we try (at least we do end up with 8 columns)
+df <- df[,-(1:3)]
+hc <- findCorrelation(cor(df), cutoff = 0.985)
+df2 <- df[,-c(hc)]
+
+# scale
+df2 <- scale(df2)
+
+# train-test split
+n_rows <- nrow(df2) # 4208261
+train_ids <- sample(1:n_rows, 0.5 * n_rows)
+x_train <- df2[train_ids, ]
+x_test <- df2[-train_ids, ]
+
+# create datasets
+batch_size <- 100
+train_dataset <- tf$cast(x_train, tf$float32) %>%
+  tensor_slices_dataset %>%
+  dataset_batch(batch_size)
+
+test_dataset <- tf$cast(x_test, tf$float32) %>%
+  tensor_slices_dataset %>%
+  dataset_batch(nrow(x_test))
+```
+
+To construct the flow, the first thing needed is the base distribution.
+
+```{r}
+base_dist <- tfd_multivariate_normal_diag(loc = rep(0, ncol(df2)))
+```
+
+Now for the flow, by default constructed with batchnorm and permutation of feature order. 
+
+```{r}
+num_hidden <- 100
+dim <- ncol(df2)
+
+use_batchnorm <- TRUE
+use_permute <- TRUE
+num_mafs <-10
+num_layers <- 3 * num_mafs
+
+bijectors <- vector(mode = "list", length = num_layers)
+
+for (i in seq(1, num_layers, by = 3)) {
+  maf <- tfb_masked_autoregressive_flow(
+    shift_and_log_scale_fn = tfb_masked_autoregressive_default_template(
+      hidden_layers = list(num_hidden, num_hidden),
+      activation = tf$nn$tanh))
+  bijectors[[i]] <- maf
+  if (use_batchnorm)
+    bijectors[[i + 1]] <- tfb_batch_normalization()
+  if (use_permute)
+    bijectors[[i + 2]] <- tfb_permute((ncol(df2) - 1):0)
+}
+
+if (use_permute) bijectors <- bijectors[-num_layers]
+
+flow <- bijectors %>%
+  discard(is.null) %>%
+  rev() %>%
+  tfb_chain()
+
+target_dist <- tfd_transformed_distribution(
+  distribution = base_dist,
+  bijector = flow
+)
+```
+
+And configuring the optimizer:
+
+```{r}
+optimizer <- tf$train$AdamOptimizer(1e-4)
+```
+
+Under that isotropic Gaussian we chose as a base distribution, how likely are the data?
+
+```{r}
+base_loglik <- base_dist %>% 
+  tfd_log_prob(x_train) %>% 
+  tf$reduce_mean()
+base_loglik %>% as.numeric()        # -11.33871
+
+base_loglik_test <- base_dist %>% 
+  tfd_log_prob(x_test) %>% 
+  tf$reduce_mean()
+base_loglik_test %>% as.numeric()   # -11.36431
+
+```
+
+
+And, just as a quick sanity check: What is the loglikelihood of the data under the transformed distribution _before any training_?
+
+```{r}
+target_loglik_pre <-
+  target_dist %>% tfd_log_prob(x_train) %>% tf$reduce_mean()
+target_loglik_pre %>% as.numeric()        # -11.22097
+
+target_loglik_pre_test <-
+  target_dist %>% tfd_log_prob(x_test) %>% tf$reduce_mean()
+target_loglik_pre_test %>% as.numeric()   # -11.36431
+```
+
+The values match - good. Here now is the training loop. Being impatient, here already we keep checking the loglikelihood on the (complete) test set to see if we're making any progress.
```

Comment:
I'd take out the second here (or maybe the first) — whichever feels more workable.